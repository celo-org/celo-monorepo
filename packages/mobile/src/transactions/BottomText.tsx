import colors from '@celo/react-components/styles/colors.v2'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import React from 'react'
import { StyleSheet, Text } from 'react-native'

interface BottomTextProps {
  children: React.ReactNode
}

export default function BottomText({ children }: BottomTextProps) {
  return <Text style={styles.text}>{children}</Text>
}

const styles = StyleSheet.create({
  text: {
    ...fontStyles.regular,
    color: colors.gray4,
    textAlign: 'center',
    marginTop: 'auto',
  },
})
