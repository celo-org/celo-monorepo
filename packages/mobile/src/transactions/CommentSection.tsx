import HorizontalLine from '@celo/react-components/components/HorizontalLine'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { Namespaces } from 'src/i18n'

interface Props {
  comment?: string | null
}

export default function CommentSection({ comment }: Props) {
  const { t } = useTranslation(Namespaces.sendFlow7)
  if (!comment) {
    return null
  }

  return (
    // Uses View instead of Fragment to workaround a glitch with LayoutAnimation
    <View>
      <HorizontalLine />
      <Text style={styles.sectionLabel}>{t('commentLabel')}</Text>
      <Text style={styles.comment}>{comment}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  sectionLabel: {
    ...fontStyles.label,
    color: colors.gray3,
    marginBottom: 4,
  },
  comment: {
    ...fontStyles.regular,
  },
})
