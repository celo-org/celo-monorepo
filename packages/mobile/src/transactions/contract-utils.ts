import { values } from 'lodash'
import { estimateGas } from 'src/web3/utils'
import { Tx } from 'web3-core'
import { TransactionObject, TransactionReceipt } from 'web3-eth'

export type TxLogger = (event: SendTransactionLogEvent) => void

export function emptyTxLogger(_event: SendTransactionLogEvent) {
  return
}

interface TxPromiseResolvers {
  receipt: (receipt: TransactionReceipt) => void
  transactionHash: (transactionHash: string) => void
  confirmation: (confirmation: boolean) => void
}

type PromiseRejection = (error: Error) => void
interface TxPromiseReject {
  receipt: PromiseRejection
  transactionHash: PromiseRejection
  confirmation: PromiseRejection
}

export interface TxPromises {
  receipt: Promise<TransactionReceipt>
  transactionHash: Promise<string>
  confirmation: Promise<boolean>
}

export function awaitConfirmation(txPromises: TxPromises) {
  return txPromises.confirmation
}

// Couldn't figure out how to make it generic
export type SendTransaction<T> = (
  tx: TransactionObject<any>,
  account: string,
  txId?: string
) => Promise<T>

export enum SendTransactionLogEventType {
  Started,
  EstimatedGas,
  ReceiptReceived,
  TransactionHashReceived,
  Confirmed,
  Failed,
  Exception,
}

interface Started {
  type: SendTransactionLogEventType.Started
}
const Started: Started = { type: SendTransactionLogEventType.Started }

interface Confirmed {
  type: SendTransactionLogEventType.Confirmed
}
const Confirmed: Confirmed = { type: SendTransactionLogEventType.Confirmed }

export type SendTransactionLogEvent =
  | Started
  | EstimatedGas
  | ReceiptReceived
  | TransactionHashReceived
  | Confirmed
  | Failed
  | Exception

interface EstimatedGas {
  type: SendTransactionLogEventType.EstimatedGas
  gas: number
}

function EstimatedGas(gas: number): EstimatedGas {
  return { type: SendTransactionLogEventType.EstimatedGas, gas }
}

interface ReceiptReceived {
  type: SendTransactionLogEventType.ReceiptReceived
  receipt: TransactionReceipt
}

function ReceiptReceived(receipt: TransactionReceipt): ReceiptReceived {
  return { type: SendTransactionLogEventType.ReceiptReceived, receipt }
}

interface TransactionHashReceived {
  type: SendTransactionLogEventType.TransactionHashReceived
  hash: string
}

function TransactionHashReceived(hash: string): TransactionHashReceived {
  return { type: SendTransactionLogEventType.TransactionHashReceived, hash }
}

interface Failed {
  type: SendTransactionLogEventType.Failed
  error: Error
}

function Failed(error: Error): Failed {
  return { type: SendTransactionLogEventType.Failed, error }
}

interface Exception {
  type: SendTransactionLogEventType.Exception
  error: Error
}

function Exception(error: Error): Exception {
  return { type: SendTransactionLogEventType.Exception, error }
}

/**
 * sendTransactionAsync mainly abstracts the sending of a transaction in a promise like
 * interface. Use the higher-order sendTransactionFactory as a consumer to configure
 * logging and promise resolution
 * TODO: Should probably renamed to sendTransaction once we remove the current
 *       sendTransaction
 * @param tx The transaction object itself
 * @param account The address from which the transaction should be sent
 * @param feeCurrencyContract The contract instance of the Token in which to pay gas for
 * @param logger An object whose log level functions can be passed a function to pass
 *               a transaction ID
 */
export async function sendTransactionAsync<T>(
  tx: TransactionObject<T>,
  account: string,
  feeCurrencyAddress: string,
  logger: TxLogger = emptyTxLogger,
  estimatedGas?: number,
  gasPrice?: string,
  nonce?: number
): Promise<TxPromises> {
  // @ts-ignore
  const resolvers: TxPromiseResolvers = {}
  // @ts-ignore
  const rejectors: TxPromiseReject = {}

  const receipt: Promise<TransactionReceipt> = new Promise((resolve, reject) => {
    resolvers.receipt = resolve
    rejectors.receipt = reject
  })

  const transactionHash: Promise<string> = new Promise((resolve, reject) => {
    resolvers.transactionHash = resolve
    rejectors.transactionHash = reject
  })

  const confirmation: Promise<boolean> = new Promise((resolve, reject) => {
    resolvers.confirmation = resolve
    rejectors.confirmation = reject
  })

  const rejectAll = (error: Error) => {
    values(rejectors).map((reject) => {
      // @ts-ignore
      reject(error)
    })
  }

  try {
    logger(Started)
    const txParams: Tx = {
      from: account,
      feeCurrency: feeCurrencyAddress,
      // Hack to prevent web3 from adding the suggested gold gas price, allowing geth to add
      // the suggested price in the selected feeCurrency.
      gasPrice: gasPrice ? gasPrice : '0',
      nonce,
    }

    if (estimatedGas === undefined) {
      estimatedGas = (await estimateGas(tx, txParams)).toNumber()
      logger(EstimatedGas(estimatedGas))
    }

    tx.send({ ...txParams, gas: estimatedGas })
      // @ts-ignore
      .on('receipt', (r: TransactionReceipt) => {
        logger(ReceiptReceived(r))
        if (resolvers.receipt) {
          resolvers.receipt(r)
        }
      })
      .on('transactionHash', (txHash: string) => {
        logger(TransactionHashReceived(txHash))

        if (resolvers.transactionHash) {
          resolvers.transactionHash(txHash)
        }
      })
      .on('confirmation', (confirmationNumber: number) => {
        if (confirmationNumber > 1) {
          // "confirmation" event is called for 24 blocks.
          // if check to avoid polluting the logs and trying to remove the standby notification more than once
          return
        }
        logger(Confirmed)

        if (resolvers.confirmation) {
          resolvers.confirmation(true)
        }
      })
      .on('error', (error: Error) => {
        logger(Failed(error))
        rejectAll(error)
      })
  } catch (error) {
    logger(Exception(error))
    rejectAll(error)
  }

  return {
    receipt,
    transactionHash,
    confirmation,
  }
}
