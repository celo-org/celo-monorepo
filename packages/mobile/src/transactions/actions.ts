import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import i18n from 'src/i18n'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { ConfirmationInput as SendConfirmationCardProps } from 'src/send/SendConfirmation'
import { TransferConfirmationCardProps } from 'src/send/TransferConfirmationCard'
import { StandbyTransaction, TransactionTypes } from 'src/transactions/reducer'
import { web3 } from 'src/web3/contracts'

export enum Actions {
  ADD_STANDBY_TRANSACTION = 'TRANSACTIONS/ADD_STANDBY_TRANSACTION',
  REMOVE_STANDBY_TRANSACTION = 'TRANSACTIONS/REMOVE_STANDBY_TRANSACTION',
  RESET_STANDBY_TRANSACTIONS = 'TRANSACTIONS/RESET_STANDBY_TRANSACTIONS',
  ADD_HASH_TO_STANDBY_TRANSACTIONS = 'TRANSACTIONS/ADD_HASH_TO_STANDBY_TRANSACTIONS',
  TRANSACTION_CONFIRMED = 'TRANSACTIONS/TRANSACTION_CONFIRMED',
}

export interface AddStandbyTransaction {
  type: Actions.ADD_STANDBY_TRANSACTION
  transaction: StandbyTransaction
}

export interface RemoveStandbyTransaction {
  type: Actions.REMOVE_STANDBY_TRANSACTION
  idx: string
}

export interface ResetStandbyTransactions {
  type: Actions.RESET_STANDBY_TRANSACTIONS
}

export interface AddHashToStandbyTransaction {
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS
  idx: string
  hash: string
}

export interface TransactionConfirmed {
  type: Actions.TRANSACTION_CONFIRMED
  txId: string
}

export type ActionTypes =
  | AddStandbyTransaction
  | RemoveStandbyTransaction
  | ResetStandbyTransactions
  | AddHashToStandbyTransaction

export const generateStandbyTransactionId = (recipientAddress: string) => {
  return web3.utils.sha3(recipientAddress + String(Date.now()))
}

export const addStandbyTransaction = (transaction: StandbyTransaction): AddStandbyTransaction => ({
  type: Actions.ADD_STANDBY_TRANSACTION,
  transaction,
})

export const removeStandbyTransaction = (idx: string): RemoveStandbyTransaction => ({
  type: Actions.REMOVE_STANDBY_TRANSACTION,
  idx,
})

export const resetStandbyTransactions = (): ResetStandbyTransactions => ({
  type: Actions.RESET_STANDBY_TRANSACTIONS,
})

export const transactionConfirmed = (txId: string): TransactionConfirmed => ({
  type: Actions.TRANSACTION_CONFIRMED,
  txId,
})

export const addHashToStandbyTransaction = (
  idx: string,
  hash: string
): AddHashToStandbyTransaction => ({
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS,
  idx,
  hash,
})

export const navigateToPaymentTransferReview = (
  type: string,
  timestamp: number,
  confirmationProps: TransferConfirmationCardProps
) => {
  let headerText = ''
  switch (type) {
    case TransactionTypes.SENT:
      headerText = i18n.t('sendFlow7:sentPayment')
      break
    case TransactionTypes.RECEIVED:
      headerText = i18n.t('receiveFlow8:receivedPayment')
      break
    case TransactionTypes.VERIFICATION_FEE:
      headerText = i18n.t('walletFlow5:verificationFee')
      break
    case TransactionTypes.FAUCET:
      headerText = i18n.t('receiveFlow8:receivedDollars')
      break
    case TransactionTypes.INVITE_SENT:
      headerText = i18n.t('inviteFlow11:inviteComplete')
      break
    case TransactionTypes.INVITE_RECEIVED:
      headerText = i18n.t('inviteFlow11:inviteReceived')
      break
    case TransactionTypes.NETWORK_FEE:
      headerText = i18n.t('walletFlow5:networkFee')
      break
  }

  navigate(Screens.TransactionReview, {
    reviewProps: {
      type,
      timestamp,
      header: headerText,
    },
    confirmationProps,
  })
}

export const navigateToExchangeReview = (
  timestamp: number,
  confirmationProps: ExchangeConfirmationCardProps
) => {
  navigate(Screens.TransactionReview, {
    reviewProps: {
      type: TransactionTypes.EXCHANGE,
      timestamp,
      header: i18n.t('exchangeFlow9:exchange'),
    },
    confirmationProps,
  })
}

export const navigateToRequestedPaymentReview = (confirmationInput: SendConfirmationCardProps) => {
  navigate(Screens.SendConfirmation, { confirmationInput })
}
