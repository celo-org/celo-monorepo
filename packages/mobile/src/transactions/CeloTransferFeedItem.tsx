// VIEW that's used on the CELO screen activity feed for CELO transfers (in or out).

import Touchable from '@celo/react-components/components/Touchable'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { CeloExchangeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { TransferItemFragment } from 'src/apollo/types'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import { formatShortenedAddress } from 'src/components/ShortenedAddress'
import { Namespaces } from 'src/i18n'
import { navigateToPaymentTransferReview } from 'src/transactions/actions'
import { TransactionStatus } from 'src/transactions/types'
import { getDatetimeDisplayString } from 'src/utils/time'

type Props = TransferItemFragment & {
  status?: TransactionStatus
}

export function CeloTransferFeedItem(props: Props) {
  const { t, i18n } = useTranslation(Namespaces.walletFlow5)
  const { address, amount, comment, status, timestamp, type } = props
  const onPress = () => {
    ValoraAnalytics.track(CeloExchangeEvents.celo_transaction_select)

    navigateToPaymentTransferReview(type, timestamp, {
      address,
      comment,
      amount,
      type,
      // fee TODO: add fee here.
    })
  }

  const dateTimeFormatted = getDatetimeDisplayString(timestamp, i18n)
  const isPending = status === TransactionStatus.Pending
  const isWithdrawal = new BigNumber(amount.value).isNegative()

  return (
    <Touchable onPress={onPress}>
      <View style={styles.container}>
        <View style={styles.firstRow}>
          <View style={styles.desc}>
            <Text style={styles.txMode}>
              {t(isWithdrawal ? 'feedItemGoldWithdrawal' : 'feedItemGoldReceived', {
                address: formatShortenedAddress(address),
              })}
            </Text>
          </View>
          <View>
            <CurrencyDisplay amount={amount} style={styles.amount} />
          </View>
        </View>
        <View style={styles.secondRow}>
          <Text style={styles.time}>{isPending ? t('confirmingExchange') : dateTimeFormatted}</Text>
        </View>
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'space-between',
    flex: 1,
    padding: variables.contentPadding,
  },
  firstRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    paddingBottom: 2,
  },
  desc: {
    flexDirection: 'row',
  },
  txMode: {
    ...fontStyles.regular500,
    color: colors.dark,
  },
  amount: {
    ...fontStyles.regular500,
    color: colors.dark,
  },
  time: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  secondRow: {},
})

export default CeloTransferFeedItem
