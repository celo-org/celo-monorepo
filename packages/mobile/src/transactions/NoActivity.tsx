import SmallButton from '@celo/react-components/components/SmallButton'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import { ApolloError } from 'apollo-boost'
import * as React from 'react'
import { withNamespaces, WithNamespaces } from 'react-i18next'
import { ActivityIndicator, Image, StyleSheet, Text, View } from 'react-native'
import { Namespaces } from 'src/i18n'
import { exchangeIcon, shinyDollar } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Stacks } from 'src/navigator/Screens'
import { FeedType } from 'src/transactions/TransactionFeed'

const goToSend = () => {
  navigate(Stacks.SendStack)
}

interface OwnProps {
  kind: FeedType
  loading: boolean
  error: ApolloError | undefined
}

type Props = OwnProps & WithNamespaces

export class NoActivity extends React.PureComponent<Props> {
  render() {
    const { kind, loading, error, t } = this.props

    if (error) {
      return (
        <View style={styles.container}>
          <View style={styles.circleRed} />
          <Text style={[fontStyles.bodySecondary, styles.text]}>{t('errorLoadingActivity.0')}</Text>
          <Text style={[fontStyles.bodySecondary, styles.text]}>{t('errorLoadingActivity.1')}</Text>
        </View>
      )
    }

    if (loading) {
      return (
        <View style={styles.container}>
          <ActivityIndicator style={styles.icon} size="large" color={colors.celoGreen} />
          <Text style={[fontStyles.bodySecondary, styles.text]}>{t('loadingActivity')}</Text>
        </View>
      )
    }

    const image = kind === FeedType.EXCHANGE ? exchangeIcon : shinyDollar
    const statusText =
      kind === FeedType.EXCHANGE ? t('noExchangeActivity') : t('noTransactionActivity')

    return (
      <View style={styles.container}>
        <Image style={styles.icon} source={image} resizeMode={'contain'} />
        <Text style={[fontStyles.bodySecondary, styles.text]}>{statusText} </Text>
        {kind === FeedType.HOME && (
          <SmallButton
            onPress={goToSend}
            text={t('global:sendCeloDollars')}
            solid={true}
            style={styles.button}
          />
        )}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    paddingHorizontal: 20,
    paddingVertical: 20,
  },
  icon: {
    marginVertical: 20,
    height: 108,
    width: 108,
  },
  text: {
    textAlign: 'center',
    marginBottom: 10,
  },
  button: {
    marginTop: 20,
    alignSelf: 'center',
  },
  circleRed: {
    marginVertical: 20,
    height: 30,
    width: 30,
    borderRadius: 15,
    backgroundColor: colors.errorRed,
  },
})

export default withNamespaces(Namespaces.walletFlow5)(NoActivity)
