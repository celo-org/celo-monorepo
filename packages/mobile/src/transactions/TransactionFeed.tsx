import ItemSeparator from '@celo/react-components/components/ItemSeparator'
import { ApolloError } from 'apollo-boost'
import * as React from 'react'
import { FlatList } from 'react-native'
import { connect } from 'react-redux'
import { Event, EventTypeNames, UserTransactionsData } from 'src/apollo/types'
import { AddressToE164NumberType } from 'src/identity/reducer'
import { Invitees, SENTINEL_INVITE_COMMENT } from 'src/invite/actions'
import { NumberToRecipient } from 'src/recipients/recipient'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'
import ExchangeFeedItem from 'src/transactions/ExchangeFeedItem'
import NoActivity from 'src/transactions/NoActivity'
import { StandbyTransaction, TransactionStatus, TransactionTypes } from 'src/transactions/reducer'
import TransferFeedItem from 'src/transactions/TransferFeedItem'
import Logger from 'src/utils/Logger'
import { privateCommentKeySelector } from 'src/web3/selectors'

const TAG = 'transactions/TransactionFeed'

export enum FeedType {
  HOME = 'home',
  EXCHANGE = 'exchange',
}

interface StateProps {
  invitees: Invitees
  commentKey: string | null | undefined
  addressToE164Number: AddressToE164NumberType
  standbyTransactions: StandbyTransaction[]
  recipientCache: NumberToRecipient
}

type Props = {
  kind: FeedType
  loading: boolean
  error: ApolloError | undefined
  data: UserTransactionsData | undefined
  standbyTransactions: StandbyTransaction[]
  standbyTransactionFilter?: (tx: StandbyTransaction) => boolean
} & StateProps

const mapStateToProps = (state: RootState): StateProps => ({
  invitees: state.invite.invitees,
  commentKey: privateCommentKeySelector(state),
  addressToE164Number: state.identity.addressToE164Number,
  standbyTransactions: state.transactions.standbyTransactions,
  recipientCache: recipientCacheSelector(state),
})

export class TransactionFeed extends React.PureComponent<Props> {
  // TODO(cmcewen): Clean this up. Standby txs should have the same data shape
  renderItem = (commentKeyBuffer: Buffer | null) => ({
    item: tx,
  }: {
    item: Event | StandbyTransaction
    index: number
  }) => {
    const { kind, addressToE164Number, invitees, recipientCache } = this.props

    if (tx.hasOwnProperty('comment')) {
      // @ts-ignore
      if (tx.comment === SENTINEL_INVITE_COMMENT) {
        if (tx.type === TransactionTypes.SENT) {
          tx.type = TransactionTypes.INVITE_SENT
        } else if (tx.type === TransactionTypes.RECEIVED) {
          tx.type = TransactionTypes.INVITE_RECEIVED
        }
      }
    }

    // @ts-ignore
    if (tx.__typename && tx.__typename === EventTypeNames.Transfer) {
      return (
        // @ts-ignore
        <TransferFeedItem
          status={TransactionStatus.Complete}
          invitees={invitees}
          addressToE164Number={addressToE164Number}
          recipientCache={recipientCache}
          commentKey={commentKeyBuffer}
          {...tx}
        />
      )
      // @ts-ignore
    } else if (tx.__typename && tx.__typename === EventTypeNames.Exchange) {
      return (
        // @ts-ignore
        <ExchangeFeedItem
          showImage={kind === FeedType.HOME}
          status={TransactionStatus.Complete}
          {...tx}
        />
      )
    } else if (tx.type && tx.type === TransactionTypes.EXCHANGE) {
      // @ts-ignore
      return <ExchangeFeedItem showImage={kind === FeedType.HOME} {...tx} />
    } else if (tx.type) {
      return (
        // @ts-ignore
        <TransferFeedItem
          recipientCache={recipientCache}
          addressToE164Number={addressToE164Number}
          invitees={invitees}
          commentKey={commentKeyBuffer}
          {...tx}
        />
      )
    } else {
      return <React.Fragment />
    }
  }

  keyExtractor = (item: Event | StandbyTransaction) => {
    return item.hash + item.timestamp.toString()
  }

  render() {
    const {
      kind,
      loading,
      error,
      data,
      standbyTransactions,
      standbyTransactionFilter,
      commentKey,
    } = this.props

    if (error) {
      Logger.error(TAG, 'Failure while loading transaction feed', error)
      return <NoActivity kind={kind} loading={loading} error={error} />
    }

    const events = (data && data.events) || []
    const commentKeyBuffer = commentKey ? Buffer.from(commentKey, 'hex') : null

    const queryDataTxIDs = new Set(events.map((event: Event) => event.hash))
    const notInQueryTxs = (tx: StandbyTransaction) =>
      !queryDataTxIDs.has(tx.id) && tx.status !== TransactionStatus.Failed
    let filteredStandbyTxs = standbyTransactions.filter(notInQueryTxs)

    if (standbyTransactionFilter) {
      filteredStandbyTxs = filteredStandbyTxs.filter(standbyTransactionFilter)
    }

    // TODO move filter to gql
    const exchangeFilter = (tx: Event) =>
      tx !== null && (kind === FeedType.EXCHANGE ? tx.__typename === EventTypeNames.Exchange : true)
    const filteredQueryTxs = events.filter(exchangeFilter)
    const txData = [...filteredStandbyTxs, ...filteredQueryTxs]

    if (txData.length > 0) {
      return (
        <FlatList
          data={txData}
          keyExtractor={this.keyExtractor}
          ItemSeparatorComponent={ItemSeparator}
          renderItem={this.renderItem(commentKeyBuffer)}
        />
      )
    } else {
      return <NoActivity kind={kind} loading={loading} error={error} />
    }
  }
}

export default connect<StateProps, {}, {}, RootState>(mapStateToProps)(TransactionFeed)
