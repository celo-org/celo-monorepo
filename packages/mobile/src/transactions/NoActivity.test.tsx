import * as React from 'react'
import 'react-native'
import * as renderer from 'react-test-renderer'
import NoActivity from 'src/transactions/NoActivity'
import { FeedType } from 'src/transactions/TransactionFeed'

it('renders loading', () => {
  const tree = renderer.create(<NoActivity loading={true} kind={FeedType.HOME} error={undefined} />)
  expect(tree).toMatchSnapshot()
})

it('renders exchange', () => {
  const tree = renderer.create(
    <NoActivity loading={false} kind={FeedType.EXCHANGE} error={undefined} />
  )
  expect(tree).toMatchSnapshot()
})

it('renders home', () => {
  const tree = renderer.create(
    <NoActivity loading={false} kind={FeedType.HOME} error={undefined} />
  )
  expect(tree).toMatchSnapshot()
})
