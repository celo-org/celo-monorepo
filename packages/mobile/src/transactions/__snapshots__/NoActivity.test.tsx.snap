// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders exchange 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "paddingHorizontal": 20,
      "paddingVertical": 20,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../packages/mobile/src/images/exchange.png",
      }
    }
    style={
      Object {
        "height": 108,
        "marginVertical": 20,
        "width": 108,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 16,
          "lineHeight": 24,
        },
        Object {
          "marginBottom": 10,
          "textAlign": "center",
        },
      ]
    }
  >
    noExchangeActivity
     
  </Text>
</View>
`;

exports[`renders home 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "paddingHorizontal": 20,
      "paddingVertical": 20,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../packages/mobile/src/images/shiny-dollar.png",
      }
    }
    style={
      Object {
        "height": 108,
        "marginVertical": 20,
        "width": 108,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 16,
          "lineHeight": 24,
        },
        Object {
          "marginBottom": 10,
          "textAlign": "center",
        },
      ]
    }
  >
    noTransactionActivity
     
  </Text>
  <View
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 7,
        "left": 7,
        "right": 7,
        "top": 7,
      }
    }
    isTVSelectable={true}
    nativeBackgroundAndroid={
      Object {
        "attribute": "selectableItemBackground",
        "type": "ThemeAttrAndroid",
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingHorizontal": 16,
          "paddingVertical": 6,
        },
        Object {
          "backgroundColor": "#42D689",
          "paddingHorizontal": 18,
          "paddingVertical": 8,
        },
        Object {
          "alignSelf": "center",
          "marginTop": 20,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#81868B",
            "fontFamily": "Hind-SemiBold",
            "fontSize": 13,
            "lineHeight": 17,
          },
          Object {
            "lineHeight": 20,
            "textAlign": "center",
          },
          Object {
            "color": "#FFFFFF",
          },
          null,
          undefined,
        ]
      }
    >
      global:sendCeloDollars
    </Text>
  </View>
</View>
`;

exports[`renders loading 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "paddingHorizontal": 20,
      "paddingVertical": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#42D689"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 108,
        "marginVertical": 20,
        "width": 108,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 16,
          "lineHeight": 24,
        },
        Object {
          "marginBottom": 10,
          "textAlign": "center",
        },
      ]
    }
  >
    loadingActivity
  </Text>
</View>
`;
